patroni/postgresql.py

```diff
@@ -90,39 +107,58 @@ def __init__(self, config):
             self._write_postgresql_conf()  # we are "joining" already running postgres
 
-    @staticmethod
-    def get_server_parameters(config):
-        return {p: v for p, v in (config.get('parameters') or {}).items() if p not in ('listen_addresses', 'port')}
+    @property
+    def use_slots(self):
+        return self._use_slots and self._major_version >= 9.4
+
+    def get_major_version(self):
+        if not self.data_directory_empty():
+            try:
+                with open(os.path.join(self._data_dir, 'PG_VERSION')) as f:
+                    return float(f.read())
+            except Exception:
+                logger.exception('Failed to read PG_VERSION from %s', self._data_dir)
+        return 0.0
+
+    def get_server_parameters(self, config):
+        parameters = config['parameters'].copy()
+        listen_addresses, port = (config['listen'] + ':5432').split(':')[:2]
+        parameters.update({'listen_addresses': listen_addresses, 'port': port})
+        return parameters
 
     def resolve_connection_addresses(self):
-        self.local_address = self.get_local_address()
+        self._local_address = self.get_local_address()
         self.connection_string = 'postgres://{username}:{password}@{connect_address}/postgres'.format(
-            connect_address=self._connect_address or self.local_address, **self.replication)
+            connect_address=self._connect_address or self._local_address, **self._replication)
 
     def reload_config(self, config):
         server_parameters = self.get_server_parameters(config)
-        self._connect_address = config.get('connect_address')
-        listen_addresses, port = (config['listen'] + ':5432').split(':')[:2]
-        if self._listen_addresses == listen_addresses and self._port == port:
-            self.resolve_connection_addresses()
 
-        self._listen_addresses = listen_addresses
-        self._port = port
+        listen_address_changed = reload_pending = False
         if self.is_healthy():
             changes = server_parameters.copy()
             changes.update({p: None for p, v in self._server_parameters.items() if p not in server_parameters})
             if changes:
-                for r in self.query("""SELECT name, setting
+                for r in self.query("""SELECT name, setting, context
                                          FROM pg_settings
-                                        WHERE context in ('internal', 'postmaster')
-                                          AND name IN (""" + ', '.join('%s' for _ in changes.keys()) + ')',
+                                        WHERE name IN (""" + ', '.join('%s' for _ in changes.keys()) + ')',
                                     *(list(changes.keys()))):
                     if server_parameters[r[0]] is None or str(server_parameters[r[0]]) != str(r[1]):
-                        self._restart_pending = True
-                        break
+                        reload_pending = True
```

Comment:
reload_pending will always be set even for those parameters that are pending restart.